### PR TITLE
[TTAHUB-1530] Make certain goal templates selectable from the AR

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change
This PR inserts via migration a goal template that will be selectable from the user interface. When selected, it will trigger another question being added to the goal form. That question has dynamic validation and text, also inserted via migration.

## How to test
- Select the goal "(FEI) The recipient will eliminate and/or reduce underenrollment as part of the Full Enrollment Initiative (as measured by monthly reported enrollment)" on a new activity report
- It should be available under any recipient.
- When you select that goal, confirm that a conditional question appears and that you can only select a maximum two answers to that question.
- Save the report, and return, and confirm that the goal appears as expected
- Approve the report and visit the approved report page and the RTR and confirm that that goal a) appears correctly on the approved report and b) appears on the RTR for that recipient

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1530

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: frontend/src/pages/ActivityReport/Pages/components/GoalPicker.js [hunk 1]
```diff
@@ -85,8 +85,12 @@ const GoalPicker = ({
...
 
   // We need options with the number and also we need to add the
-  // "create new goal to the front of all the options"
+  // goal templates and "create new goal" to the front of all the options
   const options = [
+    ...availableGoals.goalTemplates.map((g) => ({
+      ...g,
+      grantIds,
+    })),
     newGoal(grantIds),
```
Comment (GarrettEHill) [inline frontend/src/pages/ActivityReport/Pages/components/GoalPicker.js:94]: What order should this be in? Create, Template, existing or template, create, existing

Comment (thewatermethod) [reply frontend/src/pages/ActivityReport/Pages/components/GoalPicker.js:94]: I had it 
1) Template
2) Create new
3) Existing

But maybe that was taking their instructions too literally... I can follow up before we release

File: src/migrations/20230323000000-fei-template-goal.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/* eslint-disable max-len */
```
Comment (thewatermethod) [inline src/migrations/20230323000000-fei-template-goal.js:1]: note that this also borks the goal seeder if you use the test-db shell script, as the migrations run first and creates a goal template with an ID of 1, the same ID which is hardcoded into the goal seeder

Comment (thewatermethod) [reply src/migrations/20230323000000-fei-template-goal.js:1]: I guess more importantly than that, it also borks the CI :)

Comment (GarrettEHill) [reply src/migrations/20230323000000-fei-template-goal.js:1]: hmmmm ok I will need to figure that out

File: src/services/goalTemplates.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  isRttapa: string;
+  status: string;
+  endDate: string | null;
+  grantIds: [];
+  oldGrantIds: [];
+  onApprovedAR: true;
+  isNew: false;
+}
+
+export async function getCuratedTemplates(grantIds: number[] | null): Promise<GoalTemplate[]> {
```
[the remark this answers is not in the document]
Comment (GarrettEHill) [reply src/services/goalTemplates.ts:26]: I agree, I will add the bypass

File: frontend/src/pages/ActivityReport/Pages/components/GoalPicker.js (hunk 1 above)
Comment (thewatermethod) [reply frontend/src/pages/ActivityReport/Pages/components/GoalPicker.js:94]: Sounds like we have a decision, will make this change now

**Comment by thewatermethod** [comment]
@GarrettEHill lmk when this is ready for me to attack

File: src/services/goalTemplates.test.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { setFieldPromptForCuratedTemplate, setFieldPromptsForCuratedTemplate } from './goalTemplates';
+
+jest.mock('./goalTemplates', () => ({
```
Comment (thewatermethod) [inline src/services/goalTemplates.test.js:3]: Not sure if this test is meaningful as is

File: src/services/goalTemplates.ts [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+  // Retrieve the current responses and prompt requirements for the given goals and prompt ID.
+  const [currentResponses, promptRequirements] = await Promise.all([
+    GoalModel.findAll({
+      attributes: [
+        [sequelize.col('Goal.id'), 'goalId'],
+        [sequelize.col('prompts.id'), 'promptId'],
+        [sequelize.col('prompts.ordinal'), 'ordinal'],
+        [sequelize.col('responses.response'), 'response'],
+        [
+          sequelize.literal(`COALESCE("GoalFieldResponseModel"."response",array[]::text[]) != ARRAY[${response
```
Comment (thewatermethod) [inline src/services/goalTemplates.ts:215]: GoalFieldResponseModel won't be anything sequelize understands, I don't think

File: src/services/goalTemplates.ts [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+              .queryGenerator
+              .escape(r))
+            .join(',')}]::text[]`),
+          'isChanged',
+        ],
+      ],
+      where: {
+        id: goalIds,
+        [Op.or]: [
+          { '$"prompts"."id"$': '$"responses"."goalTemplateFieldPromptId"$' },
```
Comment (thewatermethod) [inline src/services/goalTemplates.ts:227]: This line generates a syntax error

Here is the SQL currently being generated:
```sql
SELECT "Goal"."id" AS "_0",
	"prompts"."id" AS "_1",
	"prompts"."ordinal" AS "_2",
	"responses"."response" AS "_3",
	"GoalFieldResponseModel"."response" != ARRAY['Community Options'] AS "_4",
	"prompts->GoalFieldResponse"."id" AS "_5",
	"prompts->GoalFieldResponse"."goalId" AS "_6",
	"prompts->GoalFieldResponse"."goalTemplateFieldPromptId" AS "_7",
	"prompts->GoalFieldResponse"."response" AS "_8",
	"prompts->GoalFieldResponse"."onAR" AS "_9",
	"prompts->GoalFieldResponse"."onApprovedAR" AS "_10",
	"prompts->GoalFieldResponse"."createdAt" AS "_11",
	"prompts->GoalFieldResponse"."updatedAt" AS "_12"
FROM "Goals" AS "Goal"
LEFT OUTER JOIN "GoalFieldResponses" AS "responses" ON "Goal"."id" = "responses"."goalId"
AND "responses"."goalTemplateFieldPromptId" = 1
INNER JOIN ("GoalFieldResponses" AS "prompts->GoalFieldResponse"
INNER JOIN "GoalTemplateFieldPrompts" AS "prompts" ON "prompts"."id" = "prompts->GoalFieldResponse"."goalTemplateFieldPromptId") ON "Goal"."id" = "prompts->GoalFieldResponse"."goalId"
AND "prompts"."id" = 1
WHERE ("prompts"."id" = '$"fieldResponses"."goalTemplateFieldPromptId"$'
	OR "prompts"."goalTemplateFieldPromptId" IS NULL)
	AND "Goal"."id" IN (37499)
```

**Review by AdamAdHocTeam — COMMENTED** [review]
Question, can you only select 1 FEI goal per AR?

The second goal I attempted to add is missing the FEI option.

**Review by AdamAdHocTeam — COMMENTED** [review]
Minor, default props validation:

image

**Review by AdamAdHocTeam — COMMENTED** [review]
Should we add a colum for FEI root cause here?

image

**Review by AdamAdHocTeam — COMMENTED** [review]
Should we add FEI root cause to CSV export?

**Comment by thewatermethod** [comment]
> Should we add FEI root cause to CSV export?
> Should we add a colum for FEI root cause here?

I think in the case of both of those, yes, but let's do that in separate PRs. No rush to do it here as the goals will only *start* being created on Friday, and adding the column to the table should go through design

**Review by kryswisnaskas — COMMENTED** [review]
Looks good. I was able to create the goal with no more than 2 reasons (without resources). Removing, editing, submitting for an approval and approving all worked as advertised. 

The only thing I noticed was the inconsistent validation messaging.  This might not necessarily be for this PR, but on "Save Draft", when more than 2 reasons were selected, we are getting "Network error in the banner", while "Save Goal" with more than 2 reasons shows everything everything is fine.

**Review by AdamAdHocTeam — APPROVED** [review]
LGTM, the issue has been fixed.

**Pull request merged**